qiskit/primitives/backend_estimator_dev.py

```diff
@@ -0,0 +1,612 @@
+            transpiled_circuit = original_circuit
+        else:
+            transpile_options = {**self.transpile_options.__dict__}
+            transpiled_circuit = transpile(original_circuit, self.backend, **transpile_options)
+        final_layout = self._infer_final_layout(original_circuit, transpiled_circuit)
+        transpiled_circuit.remove_final_measurements()
+        if transpiled_circuit.metadata is None:  # TODO: QuantumCircuit class should return {}
+            transpiled_circuit.metadata = {}
+        transpiled_circuit.metadata.update({"final_layout": final_layout})
+        return transpiled_circuit
+
+    @classmethod  # TODO: multiple registers
+    def _infer_final_layout(
+        cls, original_circuit: QuantumCircuit, transpiled_circuit: QuantumCircuit
+    ) -> Layout:
+        """Retrieve final layout from original and transpiled circuits (all measured)."""
+        physical_qubits = cls._generate_final_layout_intlist(original_circuit, transpiled_circuit)
+        layout_dict: dict[int, Any] = dict.fromkeys(range(transpiled_circuit.num_qubits))
+        for physical_qubit, virtual_qubit in zip(physical_qubits, original_circuit.qubits):
+            layout_dict.update({physical_qubit: virtual_qubit})
+        return Layout(layout_dict)
+
+    @staticmethod
+    def _generate_final_layout_intlist(
+        original_circuit: QuantumCircuit, transpiled_circuit: QuantumCircuit
+    ) -> Iterator[int]:
+        """Generate final layout intlist of physical qubits.
+
+        Note: Works under the assumption that the original circuit has a `measure_all`
+        instruction at its end, and that the transpiler does not affect the classical
+        registers.
+        """
+        # TODO: raise error if assumption in docstring is not met
+        qubit_index_map = {qubit: i for i, qubit in enumerate(transpiled_circuit.qubits)}
+        num_measurements: int = original_circuit.num_qubits
+        for i in range(-num_measurements, 0):
+            _, qargs, _ = transpiled_circuit[i]
+            physical_qubit = qargs[0]
+            yield qubit_index_map[physical_qubit]
+
+    def _run_bound_pass_manager(self, circuit: QuantumCircuit) -> QuantumCircuit:
+        """Run bound pass manager if set."""
+        if not isinstance(circuit, QuantumCircuit):
+            raise TypeError(f"Expected `QuantumCircuit`, received {type(circuit)} instead.")
+        if self._bound_pass_manager is not None:
+            circuit = self._bound_pass_manager.run(circuit)
+        return circuit
+
+    ################################################################################
+    ## MEASUREMENT
+    ################################################################################
+    @property
+    def _observable_decomposer(self) -> ObservableDecomposer:
+        """Observable decomposer based on object's config."""
+        if self.abelian_grouping:
+            return AbelianDecomposer()
+        return NaiveDecomposer()
```

Comment:
Creating a new class every time is fairly surprising behaviour for a `property` (it makes it unusually costly to query).  I'm guessing it's done here because `abelian_grouping` is exposed as explicitly mutable state, but we'd still avoid these construction costs if `decompose` and `build_pauli` were just inlined into methods on this same class that dispatched on `abelian_grouping` in the same way.

The extra abstractions in this class makes it difficult to spot that this actually causes `n_circuits * (n_observables + 1)` individual decomposer classes to be instantiated on every call to `_compute`.  This is the kind of thing I'm talking about when I say that too many properties kill us by a thousand papercuts; you'd always think that the expression `self._observable_decomposer` should be fine to call in a loop (as is done here), but we're actually allocating a whole new class every time it happens.  Couple that with the `_observable_decomposer` property triggering a second internal property (`abelian_grouping`) on every lookup as well, and you can see how this stuff adds up.  (I apply similar logic in my opposition of too many validation checks of type-hinted attributes that are supposed to be mutated / constructed quickly in Python too.) 

Of course, removing these particular instances is unlikely to have any massive impact on performance in this case, but it's the general principle that this kind of abstraction-heavy code weighs things down, and it's easily avoidable.

Reply:
Do you have any metrics on how costly it is to instantiate a new class, and what is the performance gain by not doing it? 

I ask because this saves memory by not having to maintain one of these objects in the instance dictionary, it reduces issues with mutability as you mention and some other miner things. So it would be interested to know what is the tradeoff. 

Also, do you have data on this slow-down that properties cause as apposed to plain attributes? I'd like to address this on that footing rather than in the abstract.